examples/vector_database/scripts/compute_vectors.py

```diff
@@ -0,0 +1,240 @@
+
+        if self.start_idx > 0:
+            dataset = dataset.skip(self.start_idx)
+
+        for idx, item in enumerate(dataset, start=self.start_idx):
+            if self.end_idx and idx >= self.end_idx:
+                break
+            yield idx, item
+
+    async def do_data_loading(
+            self) -> AsyncIterator[Tuple[int, Tuple[torch.Tensor, Any]]]:
+        """Load and preprocess ImageNet images."""
+        if self.model is None:
+            await self.setup_model()
+
+        async for idx, item in self.get_dataset_iterator():
+            try:
+                # ImageNet provides PIL Images directly
+                tensor = self.preprocess(item['image'])
+                if tensor is not None:
+                    # Pass through both the tensor and original image
+                    yield idx, (tensor, item['image'])
+            except Exception as e:
+                logging.debug(
+                    f"Error preprocessing image at index {idx}: {str(e)}")
+
+    async def do_batch_processing(
+        self, batch: List[Tuple[int, Tuple[torch.Tensor, Any]]]
+    ) -> List[Tuple[int, bytes]]:
+        """Process a batch of images through CLIP."""
+        if self.model is None:
+            await self.setup_model()
+
+        # Unpack the batch
+        indices, batch_data = zip(*batch)
+        model_inputs, original_images = zip(*batch_data)
+
+        # Stack inputs into a batch
+        batch_tensor = torch.stack(model_inputs).to(self.device)
+
+        # Run inference
+        with torch.no_grad():
+            features = self.model.encode_image(batch_tensor)
+            features /= features.norm(dim=-1, keepdim=True)
+
+        # Convert to numpy arrays
+        embeddings = features.cpu().numpy()
+
+        # Convert original images to base64
+        images_base64 = {}
+        for idx, img in zip(indices, original_images):
+            buffered = BytesIO()
+            img.save(buffered, format="JPEG")
+            img_str = base64.b64encode(buffered.getvalue()).decode()
+            images_base64[idx] = img_str
+
+        # Return both embeddings and images
+        return [(idx, pickle.dumps((images_base64[idx], arr)))
+                for idx, arr in zip(indices, embeddings)]
+
+    async def run(self):
+        """
+        Run the batch processing pipeline without appending to existing files.
+        Each partition is written atomically using a temporary file.
+        """
+        results = []
+        partition_counter = 0
+
+        async for idx, input_data in self.do_data_loading():
+            self._current_batch.append((idx, input_data))
+            if len(self._current_batch) >= self.batch_size:
+                batch_results = await self.do_batch_processing(
+                    self._current_batch)
+                results.extend(batch_results)
+                self._current_batch = []
+
+                if len(results) >= self.checkpoint_size:
```

Comment:
I wonder if we should just checkpoint every batch for simplicity, so we don't have to worry about handling batch_size and checkpoint_size separately.

Reply:
Oh I thought we had batch_size = 50 and checkpoint_size = 100 but I see they are both overridden in yaml